File: src/main/java/bilibili/lingaly/tutorial/item/advanced/Probe.java
package bilibili.lingaly.tutorial.item.advanced;

import bilibili.lingaly.tutorial.ModTags;
import net.fabricmc.fabric.api.item.v1.FabricItemSettings;
import net.minecraft.core.BlockPos;
import net.minecraft.network.chat.Component;
import net.minecraft.network.chat.TextComponent;
import net.minecraft.network.chat.TranslatableComponent;
import net.minecraft.world.InteractionResult;
import net.minecraft.world.entity.player.Player;
import net.minecraft.world.item.Item;
import net.minecraft.world.item.ItemStack;
import net.minecraft.world.item.TooltipFlag;
import net.minecraft.world.item.context.UseOnContext;
import net.minecraft.world.level.Level;
import net.minecraft.world.level.block.state.BlockState;
import org.jetbrains.annotations.Nullable;

import java.util.List;


public class Probe extends Item {

    public Probe(FabricItemSettings settings) {
        super(settings);
    }

    @Override
    public void appendHoverText(ItemStack itemStack, @Nullable Level level, List<Component> list, TooltipFlag tooltipFlag) {
        list.add(new TextComponent("探测棒"));
    }

    @Override
    public InteractionResult useOn(UseOnContext useOnContext) {
        if (useOnContext.getLevel().isClientSide()){
            BlockPos blockPos = useOnContext.getClickedPos();
            boolean found = false;

            for (int i = 0; i <= blockPos.getY() + 64; i++){
                BlockState blockState = useOnContext.getLevel().getBlockState(BlockPos.ZERO.below(i));
                if (blockState.is(ModTags.Blocks.valuableBlock)){
                    found = true;
                    break;
                }
            }

            Player player = useOnContext.getPlayer();
            if (found){
                player.sendMessage(new TranslatableComponent("item.probe.useonblock", "", "了"), player.getUUID());
            } else {
                player.sendMessage(new TranslatableComponent("item.probe.useonblock", "没有", ""), player.getUUID());
            }

        }


        return super.useOn(useOnContext);
    }
}

File: src/main/kotlin/bilibili/lingaly/ktmccouse/items/advanced/Probe.kt
package bilibili.lingaly.ktmccouse.items.advanced

import bilibili.lingaly.ktmccouse.ModTags
import net.minecraft.client.item.TooltipContext
import net.minecraft.entity.player.PlayerEntity
import net.minecraft.item.Item
import net.minecraft.item.ItemStack
import net.minecraft.item.ItemUsageContext
import net.minecraft.text.LiteralText
import net.minecraft.text.Text
import net.minecraft.text.TranslatableText
import net.minecraft.util.ActionResult
import net.minecraft.world.World

class Probe(settings: Settings?) : Item(settings) {

    override fun appendTooltip(
        stack: ItemStack?,
        world: World?,
        tooltip: MutableList<Text>?,
        context: TooltipContext?
    ) {
        tooltip!!
            .add(
                LiteralText(
                    "探测棒"
                )
            )
    }

    override fun useOnBlock(
        context: ItemUsageContext?
    ): ActionResult {
        if (
            context!!
                .world
                .isClient()
        ) {
            val blockPos = context.blockPos
            var found = false

            for (
            i in 0..blockPos
                .y + 64
            ) {
                val blockState = context
                    .world
                    .getBlockState(
                        blockPos
                            .down(
                                i
                            )
                    )
                if (
                    blockState
                        .isIn(
                            ModTags
                                .Blocks
                                .probeValuableBlock
                        )
                ) {
                    found = true
                    break
                }
            }

            val playerEntity: PlayerEntity = context
                .player!!
            if (found) {
                playerEntity
                    .sendMessage(
                        TranslatableText(
                            "item.probe.useonblock",
                            "",
                            "了"
                        ),
                        false
                    )
            } else {
                playerEntity
                    .sendMessage(
                        TranslatableText(
                            "item.probe.useonblock",
                            "没有",
                            ""
                        ),
                        false
                    )
            }

            context
                .stack
                .damage(
                    1,
                    context.player
                )
                { player ->
                    player!!.sendToolBreakStatus(
                        player.activeHand
                    )
                }
        }

        return super
            .useOnBlock(context)
    }

}